pypopquiz/backends/moviepy.py

```diff
@@ -0,0 +1,112 @@
+
+            # Keep a reference to the original object that read the file.
+            # moviepy leaks process references, and hence we need to close them
+            # explicitly at the end of the run() method.
+            self.reader_ref = self.video
+
+    def trim(self, start_s: int, end_s: int) -> None:
+        """Trims a video to a given start and end time measured in seconds"""
+        self.video = self.video.subclip(start_s, end_s)
+
+    def repeat(self) -> None:
+        """Concatenates a video and audio stream with itself to make a twice as long video"""
+        self.video = moviepy.editor.concatenate_videoclips([self.video, self.video])
+
+    def fade_in_and_out(self, duration_s: int, video_length_s: int) -> None:
+        """Adds a fade-in and fade-out to/from black for the audio and video stream"""
+        self.video = self.video.fx(vfx.fadein, duration_s).\
+            fx(vfx.fadeout, duration_s).\
+            fx(afx.audio_fadein, duration_s).\
+            fx(afx.audio_fadeout, duration_s)
+
+    def scale_video(self, width: int, height: int) -> None:
+        """Scales the video and pads if necessary to the requested dimensions"""
+        self.video = self.video.fx(vfx.resize, (width, height))  # TODO: padding with black
+
+    def draw_text_in_box(
+            self, video_text: str, length: int, width: int, height: int, box_height: int, move: bool,
+            top: bool
+    ) -> None:
+        """Draws a semi-transparent box either at the top or bottom and writes text in it, optionally scrolling by"""
+        self.video = Moviepy.draw_text_in_box_on_video(
+            self.video, video_text, length, width, height, box_height, move, top
+        )
+
+    @staticmethod
+    def draw_text_in_box_on_video(
+            video: moviepy.editor.VideoFileClip, video_text: str, length: float, width: int,
+            height: int, box_height: int, move: bool, top: bool
+    ) -> moviepy.editor.CompositeVideoClip:
+        """Draws a semi-transparent box either at the top or bottom and writes text in it, optionally scrolling by"""
+        y_location = 0 if top else height - box_height
+        txt = moviepy.editor.TextClip(video_text, font='Arial', color='white', fontsize=30)
+
+        video_w, _ = video.size
+
+        # Paste the text on top of a colored bar
+        txt_col = txt.on_color(
+            size=(video_w + txt.w, box_height),  # automatic height: txt.h + 10
+            color=(0, 0, 0), pos=(video_w / 20, 'center'), col_opacity=0.6
+        )
+
+        if move:
+            txt_mov = txt_col.set_position(lambda t: (max(0, int(video_w - video_w * t / float(length))), y_location))
+        else:
+            txt_mov = txt_col
+
+        duration = video.duration
+        video = moviepy.editor.CompositeVideoClip([video, txt_mov])
+        video.duration = duration
+        return video
+
+    def add_spacer(self, text: str, duration_s: float) -> None:
```

Comment:
Perhaps you should also add this method to the base-class and add an exception-raising implementation in the ffmpeg class?

Reply:
I tried to not break the ffmpeg backend run completely (it just doesn't generate the spacers). Raising an exception is also possible, but then you need to remember to flip the spacer switch when using ffmpeg. What do you think?

Reply:
Yes, backend.py has an (empty) implementation.